autoload/crystalline.vim

```diff
@@ -334,43 +334,38 @@ function! crystalline#generate_theme(theme) abort
 
   for [l:group, l:attr] in items(a:theme)
     let l:his += [crystalline#generate_hi(l:group, l:attr)]
-
-    if !get(g:, 'crystalline_enable_sep', 0)
-      continue
-    endif
-
-    for l:group_b in g:crystalline_supported_sep[l:group]
-      let l:attr_b = get(a:theme, l:group_b, [])
-      if l:attr == []
-        continue
-      endif
-
-      let l:sep_attr = [[l:attr[0][1], l:attr_b[0][1]], [l:attr[1][1], l:attr_b[1][1]]]
-      if len(l:attr) > 2
-        let l:sep_attr += [l:attr[2]]
-      endif
-
-      if l:group ==# 'TabType'
-        let l:attr_type = has('gui_running') ? 1 : 0
-        if l:attr[l:attr_type][1] == l:attr_b[l:attr_type][1]
-          call add(g:crystalline_tab_type_fake_separators, l:group_b)
-        endif
-      endif
-
-      let l:sep_group = crystalline#get_sep_group(l:group, l:group_b)
-      let l:his += [crystalline#generate_hi(l:sep_group, l:sep_attr)]
-    endfor
   endfor
 
   if len(l:his) > 0
     exec join(l:his, ' | ')
   endif
+  let g:crystalline_theme_config = copy(a:theme)
 endfunction
 
 function! crystalline#mode_hi() abort
   return g:crystalline_mode_hi_groups[crystalline#mode_type()]
 endfunction
 
+function! crystalline#generate_sep_hi(group_a, group_b) abort
+  let l:attr_a = g:crystalline_theme_config[a:group_a]
```

Comment:
I'm not sure I'm comfortable with the level of guarantee that `g:crystalline_theme_config` will be set since `generate_sep` is called on a per-render basis and `generate_theme` is called when creating a theme, and `g:crystalline_theme_config` will never be set explicitly by `crystalline#set_theme()`. The other method you mentioned of grabbing the highlight color may be more desirable, if it is fast enough.

Reply:
I'm not sure I follow. Do you mean that even though `crystalline#set_theme()` calls into the theme's `set_theme()` function, the theme`s `set_theme()` is not guaranteed to call `crystalline#generate_theme()` (which is the place where `g:crystalline_theme_config is set)? If so, I agree – should I move setting the variable from `crystalline#generate_theme()` over to `crystalline#set_theme()`?

Reply:
Actually, `g:crystalline_theme_config` feels like a leaky implementation detail that ideally shouldn't be exposed to the user at all. I implemented it mainly as a workaround to make my PRs independent, and that was before I did any benchmarking.

With the benchmarks appearing to show that there's no difference in performance between the `synIDattrs()`-based appreach in #22 and the one based on `g:crystalline_theme_config`. So if it's alright with you, I'd like to move the `synIDattrs()`-based separator highlight group creation from #22 to this PR and re-run benchmarks.

Reply:
>the theme's set_theme() is not guaranteed to call crystalline#generate_theme()
That's exactly it. We can't guarantee a call to `generate_theme` because, well, there's nothing stopping any other plugin from creating their own theme, and at the moment all you need to do to create a theme is define some highlight groups, so there's no reason to use `generate_theme` other than convenience.

And while, right now, you'd have to be crazy to define all of the different separator highlight groups yourself, the concept behind this PR actually makes that a null issue. In fact, if separator groups are created dynamically, I might consider getting rid of the `generate_theme` function altogether.
>I'd like to try out the synIDattrs()-based separator highlight group creation and re-run benchmarks.
Sounds good to me

Reply:
> I might consider getting rid of the generate_theme function altogether.

:thinking: Huh. I hadn't even thought that far ahead, but you're not wrong, hah.

So themes would just consist of some `:hi ctermbg=bla …`?

Reply:
>So themes would just consist of some :hi ctermbg=bla …?

Yup, that's how it worked before I needed to start generating separator groups. I can't hope to replace the VIM highlight command.

Reply:
>At that point you could create Crystalline themes as actual colorscheme files, which just define some highlight groups whose names start with Crystalline

Yup, and that's how some colorschemes set themes for other statusline functions.

>So just get a list of highlight groups whose names match Crystalline.\{-}To.\{-} and run crystalline#generate_sep_hi() on them.

I think I'll have to see it - I speak human language worse than code (I write code more often)